Changes in module 'nf-core/multiqc'
--- modules/nf-core/multiqc/main.nf
+++ modules/nf-core/multiqc/main.nf
@@ -4,7 +4,7 @@
     conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/multiqc:1.21--pyhdfd78af_0' :
-        'biocontainers/multiqc:1.21--pyhdfd78af_0' }"
+        'community.wave.seqera.io/library/multiqc:1.22.1--4886de6095538010' }"
 
     input:
     path  multiqc_files, stageAs: "?/*"

************************************************************
