Changes in module 'nf-core/maltextract'
--- modules/nf-core/maltextract/main.nf
+++ modules/nf-core/maltextract/main.nf
@@ -8,19 +8,20 @@
         'biocontainers/hops:0.35--hdfd78af_1' }"
 
     input:
-    path rma6
+    tuple val(meta), path(rma6)
     path taxon_list
     path ncbi_dir
 
     output:
-    path "results"      , emit: results
-    path "versions.yml" , emit: versions
+    tuple val(meta), path("results"), emit: results
+    path "versions.yml"             , emit: versions
 
     when:
     task.ext.when == null || task.ext.when
 
     script:
     def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: 'results'
     """
     MaltExtract \\
         -Xmx${task.memory.toGiga()}g \\
@@ -28,7 +29,7 @@
         -i ${rma6.join(' ')} \\
         -t $taxon_list \\
         -r $ncbi_dir \\
-        -o results/ \\
+        -o $prefix \\
         $args
 
     cat <<-END_VERSIONS > versions.yml

************************************************************
