Changes in module 'nf-core/cutadapt'
--- modules/nf-core/cutadapt/environment.yml
+++ modules/nf-core/cutadapt/environment.yml
@@ -4,4 +4,4 @@
   - bioconda
   - defaults
 dependencies:
-  - bioconda::cutadapt=3.4
+  - bioconda::cutadapt=4.8

--- modules/nf-core/cutadapt/main.nf
+++ modules/nf-core/cutadapt/main.nf
@@ -4,14 +4,14 @@
 
     conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/cutadapt:3.4--py39h38f01e4_1' :
-        'biocontainers/cutadapt:3.4--py39h38f01e4_1' }"
+        'https://depot.galaxyproject.org/singularity/cutadapt:4.8--py39hf95cd2a_0' :
+        'biocontainers/cutadapt:4.8--py39hf95cd2a_0' }"
 
     input:
     tuple val(meta), path(reads)
 
     output:
-    tuple val(meta), path('*.trim.fastq.gz'), emit: reads
+    tuple val(meta), path('*.trimmed.fastq.gz'), emit: reads
     tuple val(meta), path('*.log')          , emit: log
     path "versions.yml"                     , emit: versions
 
@@ -21,7 +21,7 @@
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
-    def trimmed  = meta.single_end ? "-o ${prefix}.trim.fastq.gz" : "-o ${prefix}_1.trim.fastq.gz -p ${prefix}_2.trim.fastq.gz"
+    def trimmed  = meta.single_end ? "-o ${prefix}.trimmed.fastq.gz" : "-o ${prefix}_1.trimmed.fastq.gz -p ${prefix}_2.trimmed.fastq.gz"
     """
     cutadapt \\
         --cores $task.cpus \\

************************************************************
